_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,562 @@
+```
+
+The plugin enables OIDC login capability for kubectl. Let's test it out first. Run the following command, make sure to replace `k8s_oidc_issuer_url` and `k8s_oidc_client_id` with what you saved earlier during Okta set up.
+
+```bash
+kubectl oidc-login setup --oidc-issuer-url=<k8s_oidc_issuer_url> --oidc-client-id=<k8s_oidc_client_id>
+```
+
+This should produce an output that looks like this:
+
+```
+authentication in progress...
+Opening in existing browser session.
+
+## 2. Verify authentication
+
+You got a token with the following claims:
+
+{
+  "sub": "00u28lpcoheT48W5A5d7",
+  "ver": 1,
+  "iss": "<k8s_oidc_issuer_url>",
+  "aud": "<k8s_oidc_client_id>",
+  "iat": 1634657188,
+  "exp": 1634660788,
+  "jti": "ID.AmO7M7Z4xh_rSN4IHVuQEfJ4QuoADOWHSrVHuNyn9so",
+  "amr": [
+    "pwd"
+  ],
+  "idp": "0oa28lrldhsBglcC35d7",
+  "nonce": "2Rsv7E84vo_K14J6qm9IgKjNsCUxb4Hc16qgzx9CKlk",
+  "auth_time": 1634649927,
+  "at_hash": "veXQkvqTjCl70OkZnKZVyA",
+  "groups": [
+    "k8s-admins"
+  ]
+}
+
+```
+
+Now let us update kubectl configuration to add an OIDC user:
+
+```bash
+kubectl config set-credentials oidc-user \
+--exec-api-version=client.authentication.k8s.io/v1beta1 \
+--exec-command=kubectl \
+--exec-arg=oidc-login \
+--exec-arg=get-token \
+--exec-arg=--oidc-issuer-url=<k8s_oidc_issuer_url> \
+--exec-arg=--oidc-client-id=<k8s_oidc_client_id> \
+--exec-arg=--oidc-extra-scope="email offline_access profile openid"
+```
+
+Now we can use this user to authenticate to the cluster via kubectl.
+
+```bash
+kubectl get pods --user=oidc-user -n default
+```
+
+This should open a browser window to authenticate you. You can use your Okta account credentials to log in.
+
+You can play around with different users in different groups to test out the RBAC configuration we did. Users who are not in the `k8s-admins` group in Okta should not be able to access any resource. Users who are only in the `restricted-users` group should only be able to see pods and services.
+
+You can use the below command to set this user as default for your current kubectl context.
+
+```
+kubectl config set-context --current --user=oidc-user
+```
+
+# Learn more about using OIDC with Kubernetes
+
+Using OIDC is a great way to secure your Kubernetes clusters, especially in bigger teams. It is more secure than any of the default Kubernetes authentication mechanisms. On top of that, it will let you manage users and their roles and permissions in your cluster and even add multi-factor authentication for your cluster.
+
+While this exercise shows how to do it with Okta, the process would be very similar for other OIDC providers as well. Check out the [Okta Terraform provider docs](https://registry.terraform.io/providers/okta/okta/latest/docs) to see what else can be automated via Terraform. You can also automate the Kubernetes parts via terraform using the [Kubernetes Terraform provider](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs).
```

Comment:
```suggestion
While this exercise shows you an Okta implementation, the process would be very similar for other OIDC providers. Check out the [Okta Terraform provider docs](https://registry.terraform.io/providers/okta/okta/latest/docs) to see what else can be automated via Terraform. You can also automate the Kubernetes parts via terraform using the [Kubernetes Terraform provider](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs).
```
clarifies  "it" in "do it with Okta" since the context is missing in this paragraph